app/background.cc

```diff
@@ -0,0 +1,161 @@
+            last = now;
+            // display thumbnail frame with overlay info if debug enabled
+            if (pbkd->dbg) {
+                char msg[40];
+                snprintf(msg, sizeof(msg), "FPS:%0.1f FRM:%d", 1e9/nsec, pbkd->frm);
+                cv::Mat frame;
+                cv::resize(pbkd->raw, frame, cv::Size(240,160));
+                cv::putText(frame, msg, cv::Point(5,frame.rows-5), cv::FONT_HERSHEY_PLAIN, 1.0, cv::Scalar(255,255,255));
+                cv::imshow("Background", frame);
+            }
+            pbkd->mux.unlock();
+            // wait for next frame, some sources are real-time, others are not, this ensures all play in real-time.
+            if (pbkd->frm>1) {
+                long adj = (now.tv_sec-proc.tv_sec)*1000000000l + (now.tv_nsec-proc.tv_nsec);
+                now.tv_sec = 0;
+                now.tv_nsec= (long)(1.0/pbkd->fps*1e9);
+                // adjust for processing time or skip if less than that
+                if (now.tv_nsec > adj) {
+                    now.tv_nsec -= adj;
+                    nanosleep(&now, nullptr);
+                } else {
+                    if (pbkd->dbg) fprintf(stderr, "background: wait=%ld adj=%ld\n", now.tv_nsec, adj);
+                }
+                clock_gettime(CLOCK_REALTIME, &proc);
+            }
+        } else {
+            // no more frames, but if we processed some, try to reset position and go again
+            if (pbkd->frm>0 && pbkd->cap.set(cv::CAP_PROP_POS_FRAMES, 0)) {
+                std::unique_lock<std::mutex> hold(pbkd->mux);
+                pbkd->frm = 0;
+            } else {
+                // unable to reset or previous attempt produced no more frames - stop
+                if (pbkd->dbg) fprintf(stderr, "background: thread stopping at end of stream and not resettable\n");
+                break;
+            }
+        }
+    }
+    if (pbkd->dbg) fprintf(stderr, "background: thread stop\n");
+}
+
+void *load_background(const char *path, int debug) {
+    background_t *pbkd = new background_t;
+    background_t &bkd = *pbkd;
+
+    bkd.dbg = debug;
+    bkd.cap.open(path, cv::CAP_ANY);    // explicitly ask for auto-detection of backend
+    if (!bkd.cap.isOpened()) {
+        if (bkd.dbg) fprintf(stderr, "cv::VideoCapture cannot open: %s\n", path);
+        delete pbkd;
+        return nullptr;
+    }
+    bkd.cap.set(cv::CAP_PROP_CONVERT_RGB, true);
+    int fcc = (int)bkd.cap.get(cv::CAP_PROP_FOURCC);
+    bkd.fps = bkd.cap.get(cv::CAP_PROP_FPS);
+    int cnt = (int)bkd.cap.get(cv::CAP_PROP_FRAME_COUNT);
+    // Here be the logic...
+    //  if: can read 2 video frames => it's a video
+    //  else: is loaded as an image => it's an image
+    //  else: it's not usable.
+    if (bkd.cap.read(bkd.raw) && bkd.cap.read(bkd.raw)) {
+        // it's a video, try a reset and start reader thread..
+        if (bkd.cap.set(cv::CAP_PROP_POS_FRAMES, 0))
+            bkd.frm = 0;
+        else
+            bkd.frm = 2;
```

Comment:
```suggestion
        bkd.frm = bkd.cap.set(cv::CAP_PROP_POS_FRAMES, 0) ? 0 : 2;
```

Apart from this explain those magic numbers …

Reply:
I will add a suitable comment!